[feature/v0.2]  `DataTree` fails on recursion

In Python 3.11:
```
test/models/executor/version_0_2/test_models.py:22: in <module>
    from ibm_quantum_schemas.models.executor.version_0_2.models import (
ibm_quantum_schemas/models/executor/version_0_2/models.py:170: in <module>
    class QuantumProgramModel(BaseModel):
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py:237: in __new__
    complete_model_class(
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py:597: in complete_model_class
    schema = gen_schema.generate_schema(cls)
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:711: in generate_schema
    schema = self._generate_schema_inner(obj)
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1004: in _generate_schema_inner
    return self._model_schema(obj)
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:837: in _model_schema
    {k: self._generate_md_field_schema(k, v, decorators) for k, v in fields.items()},
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:837: in <dictcomp>
    {k: self._generate_md_field_schema(k, v, decorators) for k, v in fields.items()},
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1206: in _generate_md_field_schema
    common_field = self._common_field_schema(name, field_info, decorators)
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1372: in _common_field_schema
    schema = self._apply_annotations(
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2297: in _apply_annotations
    schema = get_inner_schema(source_type)
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_schema_generation_shared.py:83: in __call__
    schema = self._handler(source_type)
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2279: in inner_handler
    schema = self._generate_schema_inner(obj)
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1009: in _generate_schema_inner
    return self.match_type(obj)
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1123: in match_type
    return self._match_generic_type(obj, origin)
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1146: in _match_generic_type
    return self._union_schema(obj)
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1434: in _union_schema
    choices.append(self.generate_schema(arg))
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:711: in generate_schema
    schema = self._generate_schema_inner(obj)
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1009: in _generate_schema_inner
    return self.match_type(obj)
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1123: in match_type
    return self._match_generic_type(obj, origin)
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1150: in _match_generic_type
    return self._list_schema(self._get_first_arg_or_any(obj))
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:372: in _list_schema
    return core_schema.list_schema(self.generate_schema(items_type))
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:711: in generate_schema
    schema = self._generate_schema_inner(obj)
/root/miniforge3/envs/env5/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1009: in _generate_schema_inner
    return self.match_type(obj)
E   RecursionError: maximum recursion depth exceeded in comparison
E   If you made use of an implicit recursive type alias (e.g. `MyType = list['MyType']), consider using PEP 695 type aliases instead. For more details, refer to the documentation: https://docs.pydantic.dev/2.11/concepts/types/#named-recursive-types
!!! Recursion detected (same locals & position)
```

--- a/ibm_quantum_schemas/models/executor/version_0_2/models.py
+++ b/ibm_quantum_schemas/models/executor/version_0_2/models.py
@@ -15,7 +15,7 @@
 from __future__ import annotations
 
 import datetime
-from typing import Annotated, Literal, TypeAlias
+from typing import Annotated, Literal, TypeAlias, Union
 
 from pydantic import BaseModel, Field, model_validator
 
@@ -26,9 +26,11 @@ from ...qpy_model import QpyModelV13ToV17
 from ...samplex_model import SamplexModelSSV1ToSSV2 as SamplexModel
 from ...tensor_model import F64TensorModel, TensorModel
 
-DataTree: TypeAlias = (
-    list["DataTree"] | dict[str, "DataTree"] | TensorModel | str | float | int | bool | None
-)
+# Ignoring UP007 (auto upgrade Union to |) is required here some versions of Pydantic can't handle
+# this recursive forward reference.
+DataTree: TypeAlias = Union[  # noqa: UP007
+    list["DataTree"], dict[str, "DataTree"], TensorModel, str, float, int, bool, None
+]
 """Arbitrary nesting of lists and dicts with typed leaves."""
 
 

--- a/ibm_quantum_schemas/models/executor/version_0_2/models.py
+++ b/ibm_quantum_schemas/models/executor/version_0_2/models.py
@@ -15,7 +15,7 @@
 from __future__ import annotations
 
 import datetime
-from typing import Annotated, Literal, TypeAlias, Union
+from typing import Annotated, Literal, Union
 
 from pydantic import BaseModel, Field, model_validator
 
@@ -26,9 +26,9 @@ from ...qpy_model import QpyModelV13ToV17
 from ...samplex_model import SamplexModelSSV1ToSSV2 as SamplexModel
 from ...tensor_model import F64TensorModel, TensorModel
 
-# Ignoring UP007 (auto upgrade Union to |) is required here some versions of Pydantic can't handle
-# this recursive forward reference.
-DataTree: TypeAlias = Union[  # noqa: UP007
+# Ignoring UP007 (auto upgrade Union to |) is required here some versions of Python+Pydantic can't
+# handle this recursive forward reference.
+DataTree = Union[  # noqa: UP007
     list["DataTree"], dict[str, "DataTree"], TensorModel, str, float, int, bool, None
 ]
 """Arbitrary nesting of lists and dicts with typed leaves."""

--- a/ibm_quantum_schemas/models/executor/version_0_2/models.py
+++ b/ibm_quantum_schemas/models/executor/version_0_2/models.py
@@ -28,9 +28,10 @@ from ...samplex_model import SamplexModelSSV1ToSSV2 as SamplexModel
 from ...tensor_model import F64TensorModel, TensorModel
 
 # TypeAliasType is required for Pydantic to handle this recursive type correctly.
+# TensorModel must come before dict so Pydantic tries it first during deserialization.
 DataTree = TypeAliasType(
     "DataTree",
-    list["DataTree"] | dict[str, "DataTree"] | TensorModel | str | float | int | bool | None,
+    list["DataTree"] | TensorModel | dict[str, "DataTree"] | str | float | int | bool | None,
 )
 """Arbitrary nesting of lists and dicts with typed leaves."""
 

--- a/ibm_quantum_schemas/models/executor/version_0_2/models.py
+++ b/ibm_quantum_schemas/models/executor/version_0_2/models.py
@@ -15,7 +15,7 @@
 from __future__ import annotations
 
 import datetime
-from typing import Annotated, Literal
+from typing import Annotated, Literal, TypeAlias
 
 from pydantic import BaseModel, Field, model_validator
 
@@ -26,11 +26,14 @@ from ...qpy_model import QpyModelV13ToV17
 from ...samplex_model import SamplexModelSSV1ToSSV2 as SamplexModel
 from ...tensor_model import F64TensorModel, TensorModel
 
-DataTreeLeaf = TensorModel | str | float | int | bool | None
-"""Leaf types for a data tree structure."""
+DataTree: TypeAlias = (
+    list["DataTree"] | dict[str, "DataTree"] | TensorModel | str | float | int | bool | None
+)
+"""Arbitrary nesting of lists and dicts with typed leaves.
 
-DataTree = DataTreeLeaf | list["DataTree"] | dict[str, "DataTree"]
-"""Arbitrary nesting of lists and dicts with typed leaves."""
+Leaf values can be a standard JSON Python types (``str``, ``float``, ``int``, ``bool``, ``None``),
+or a :class:`~.TensorModel`.
+"""
 
 
 class ParamsModel(BaseParamsModel):
@@ -318,8 +321,3 @@ class QuantumProgramResultModel(BaseModel):
 
     passthrough_data: DataTree = None
     """Arbitrary nested data passed through execution without modification."""
-
-
-# Rebuild models to resolve recursive DataTree type
-QuantumProgramModel.model_rebuild()
-QuantumProgramResultModel.model_rebuild()

--- a/ibm_quantum_schemas/models/executor/version_0_2/models.py
+++ b/ibm_quantum_schemas/models/executor/version_0_2/models.py
@@ -28,6 +28,8 @@ from ...samplex_model import SamplexModelSSV1ToSSV2 as SamplexModel
 from ...tensor_model import F64TensorModel, TensorModel
 
 # TypeAliasType is required for Pydantic to handle this recursive type correctly.
+# Note that TypeAliasType is a backport for Python<3.12, so that when drop Python 3.11 support and
+# lower, this can be updated to `type DataTree = ...`.
 # TensorModel must come before dict so Pydantic tries it first during deserialization.
 DataTree = TypeAliasType(
     "DataTree",

--- a/ibm_quantum_schemas/models/executor/version_0_2/models.py
+++ b/ibm_quantum_schemas/models/executor/version_0_2/models.py
@@ -204,7 +204,7 @@ class QuantumProgramModel(BaseModel):
          over shots.
     """
 
-    passthrough_data: DataTree = None
+    passthrough_data: Annotated[DataTree, Field(default=None)]
     """Arbitrary nested data passed through execution without modification."""
 
 
@@ -317,7 +317,7 @@ class QuantumProgramResultModel(BaseModel):
     metadata: MetadataModel
     """Execution metadata pertaining to the job as a whole."""
 
-    passthrough_data: DataTree = None
+    passthrough_data: Annotated[DataTree, Field(default=None)]
     """Arbitrary nested data passed through execution without modification."""
 
 

--- a/ibm_quantum_schemas/models/executor/version_0_2/models.py
+++ b/ibm_quantum_schemas/models/executor/version_0_2/models.py
@@ -29,11 +29,7 @@ from ...tensor_model import F64TensorModel, TensorModel
 DataTree: TypeAlias = (
     list["DataTree"] | dict[str, "DataTree"] | TensorModel | str | float | int | bool | None
 )
-"""Arbitrary nesting of lists and dicts with typed leaves.
-
-Leaf values can be a standard JSON Python types (``str``, ``float``, ``int``, ``bool``, ``None``),
-or a :class:`~.TensorModel`.
-"""
+"""Arbitrary nesting of lists and dicts with typed leaves."""
 
 
 class ParamsModel(BaseParamsModel):

--- a/ibm_quantum_schemas/models/executor/version_0_2/models.py
+++ b/ibm_quantum_schemas/models/executor/version_0_2/models.py
@@ -318,3 +318,8 @@ class QuantumProgramResultModel(BaseModel):
 
     passthrough_data: DataTree = None
     """Arbitrary nested data passed through execution without modification."""
+
+
+# Rebuild models to resolve recursive DataTree type
+QuantumProgramModel.model_rebuild()
+QuantumProgramResultModel.model_rebuild()

--- a/ibm_quantum_schemas/models/executor/version_0_2/models.py
+++ b/ibm_quantum_schemas/models/executor/version_0_2/models.py
@@ -15,9 +15,10 @@
 from __future__ import annotations
 
 import datetime
-from typing import Annotated, Literal, Union
+from typing import Annotated, Literal
 
 from pydantic import BaseModel, Field, model_validator
+from typing_extensions import TypeAliasType
 
 from ....aliases import Self
 from ...base_params_model import BaseParamsModel
@@ -26,11 +27,11 @@ from ...qpy_model import QpyModelV13ToV17
 from ...samplex_model import SamplexModelSSV1ToSSV2 as SamplexModel
 from ...tensor_model import F64TensorModel, TensorModel
 
-# Ignoring UP007 (auto upgrade Union to |) is required here some versions of Python+Pydantic can't
-# handle this recursive forward reference.
-DataTree = Union[  # noqa: UP007
-    list["DataTree"], dict[str, "DataTree"], TensorModel, str, float, int, bool, None
-]
+# TypeAliasType is required for Pydantic to handle this recursive type correctly.
+DataTree = TypeAliasType(
+    "DataTree",
+    list["DataTree"] | dict[str, "DataTree"] | TensorModel | str | float | int | bool | None,
+)
 """Arbitrary nesting of lists and dicts with typed leaves."""
 
 
@@ -204,7 +205,7 @@ class QuantumProgramModel(BaseModel):
          over shots.
     """
 
-    passthrough_data: Annotated[DataTree, Field(default=None)]
+    passthrough_data: DataTree = None
     """Arbitrary nested data passed through execution without modification."""
 
 
@@ -317,9 +318,5 @@ class QuantumProgramResultModel(BaseModel):
     metadata: MetadataModel
     """Execution metadata pertaining to the job as a whole."""
 
-    passthrough_data: Annotated[DataTree, Field(default=None)]
+    passthrough_data: DataTree = None
     """Arbitrary nested data passed through execution without modification."""
-
-
-QuantumProgramModel.model_rebuild()
-QuantumProgramResultModel.model_rebuild()

--- a/ibm_quantum_schemas/models/executor/version_0_2/models.py
+++ b/ibm_quantum_schemas/models/executor/version_0_2/models.py
@@ -319,3 +319,7 @@ class QuantumProgramResultModel(BaseModel):
 
     passthrough_data: DataTree = None
     """Arbitrary nested data passed through execution without modification."""
+
+
+QuantumProgramModel.model_rebuild()
+QuantumProgramResultModel.model_rebuild()

